API/Pagination.md

```diff
@@ -0,0 +1,70 @@
+
+### Page-based
+
+Page-based pagination divides a set of records into pages of equal size (except for the last page which can have fewer records than other pages). A number is assigned to every page, starting with 1.
+
+![page-based-pagination](/img/page-based-pagination.png)
+
+The API should support client-set page size and page number params. An endpoint with pagination params applied could look like this: `/api/v1/countries?page[size]=50&page[number]=3`. In the example, the client requests the 3rd page of countries with maximum 50 records.
+This means that the API will respond with countries from 101 to 150, if the API database contains that many countries. Here are some cases when it doesn't:
+
+- 120 countries: the client would receive only 20 countries (from 101 to 120)
+- 160 countries: the client would receive 50 of them (from 101 to 150)
+
+If the client omits the page size param, you should set a default value. You should also set a maximum value (so clients can't request all records by providing an absurdly large number). If a client requests more records than the maximum allows, you can either clamp the value to the maximum or respond with an error, whatever you think is the right choice.
+
+When the page number param is omitted, you should respond with the first page. Respond with an error if the param is less than 1, but if it's greater than the total number of pages, then return the last page. Returning the last page instead of responding with an error prevents an edge case when pages disappear due to records being deleted but clients are still requesting the missing page (which they think exists).
+
+The response should optionally include pagination metadata:
+
+- current page (in case it differs from the requested page)
+- total number of pages
+- total number of items
+
+This information is optional because it won't always be used by clients, and, since it typically requires executing an additional SQL COUNT query, it can be a costly operation.
+
+Since this type of pagination allows clients to request any page, it is usually represented in the UI with a paginator where you can select a page (e.g. `<- 2 3 4 5 6 ->`). If this is the kind of UI you're implementing the pagination for, then the page-based type is the right choice.
+
+This type is also popular because it's easy to implement using SQL's `OFFSET` clause, so you may often hear it being referred to as offset-based pagination. `page[size]=50&page[number]=3` simply translates to `LIMIT 50 OFFSET 50 * (3 - 1) => LIMIT 50 OFFSET 100` in SQL. However, offset-based pagination has a negative side. The clause requires scanning _all_ rows included by the offset value, which is inefficient for large offsets and causes performance to suffer the more pages you have (e.g. page 100 is slower to fetch than page 1). If you don't have to paginate many records, then it's not that much of an issue.
```

Comment:
>  The clause requires scanning _all_ rows included by the offset value, which is inefficient for large offsets and causes performance to suffer the more pages you have

- I am not sure that this is completely true, at least in the PostgreSQL
- As far as I know, there are ways how we can arrange indexes in PostgreSQL so that they can be used for OFFSET
- Another problem is LIMIT, you need to use SORT BY for it, and also as far as I know we can use indexes so that it does not have to sort the whole table

Reply:
I'll open this as a task so we can discuss it 🙂